capreolus/index/faiss.py

```diff
@@ -0,0 +1,33 @@
+from capreolus import ConfigOption, constants, get_logger
+
+from . import Index
+
+
+logger = get_logger(__name__)
+
+
+@Index.register
+class FAISSIndex(Index):
+    module_name = "faiss"
+    
+    dependencies = [Dependency(key="encoder", module="encoder"), Dependency(key="index", module="index")] + Index.dependencies
```

Comment:
So the `encoder` is a dependency of the FAISSIndex. This would be how an ANN task invocation looks:

`python run.py task.rank with searcher.name=faiss searcher.index.encoder.name=SentenceBERT`

An alternative is to make the `encoder` the dependency of a separate `ANNRank` task:

`python run.py task.annrank with encoder.name=SentenceBERT`

Which one would you prefer? The first one is more "correct" - it is pretty clear from the invocation that the encoder is related only to the FAISS index, and sorta shows how flexible capreolus is. However, the second one is just plain easier, though a bit arbitrary.

One solution: keep both. The first one is the true invocation, and "task.annrank" would be an alias.
^ drawback: Now we have to support aliases :/

Reply:
I prefer the former (longer one), just because it allows for more flexibility with different tasks. It's nice that it works with any task that uses a Searcher, regardless of whether e.g. BM25 or FAISS is being used.